### PR TITLE
Tech: Bump TypeScript, remove Protractor and migrate to Jest

### PR DESCRIPTION
**Pull request by gaetanmaisse**
**Issue**

https://github.com/gravitee-io/issues/issues/5101

**Description**

This is the first part of a big cleanup and upgrade of this project.

_TLDR; You can review this PR commit by commit and read the details in commit messages._

I tested locally all the NPM scripts before each commit and play a bit with the app, everything looks ✅  but I would really appreciate feedback from more experimented Graviters.

Details:

## TypeScript

Bump TypeScript from 2.9.2 to 4.1.3 and fix breaking changes

## Protractor 

Protractor may have been used one day to run some E2E tests. However:
 - there is no real E2E tests
 - and moreover, it has been removed from any CI steps more than 2 years ago (in 1bb7378). Currently, we are just downloading selenium and protractor for pleasure in each CI run but we are not using them.

## Karma

Karma is the historical solution for running unit tests in an AngularJS app.
However, it is now quite outdated and unpleasant to use. Jest is another JS testing framework with both better performances, dev experience.

The command to run the unit tests is still the same, npm test, but Jest adds some useful options like:

- Running test from a single test file: npm test -- --testPathPattern=analytics.service.spec.ts
- A top-of-the-art watch mode: npm test -- --testPathPattern=analytics.service.spec.ts --watch
- A perfectly working integration with VSCode and Jetbrains IDE

As we now have running unit tests I also added a unit tests step in CI workflow.

## Dependencies

Mainly remove unused dependencies.

---

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -134,6 +134,16 @@
               <arguments>run lint</arguments>
             </configuration>
           </execution>
+          <execution>
+            <id>npm test:ci</id>
+            <goals>
+              <goal>npm</goal>
+            </goals>
+            <phase>generate-resources</phase>
```
Comment (gcusnieux) [inline pom.xml:142]: Ok maven it's overkill... but we can use test phase for this ?

File: package.json [hunk 2]
```diff
@@ -104,34 +104,11 @@
...
-    "gulp-insert": "^0.5.0",
-    "gulp-jshint": "^2.1.0",
-    "gulp-load-plugins": "^1.6.0",
-    "gulp-minify-html": "^1.0.6",
-    "gulp-ng-annotate": "^2.1.0",
-    "gulp-rename": "^1.4.0",
-    "gulp-replace": "^1.0.0",
-    "gulp-rev": "^8.1.1",
-    "gulp-rev-replace": "^0.4.4",
-    "gulp-sass": "^4.0.2",
```
Comment (gcusnieux) [inline package.json:125]: ❤️

Comment (gcusnieux) [reply package.json:125]: but gulp is used by severals script... `npm run lint`, `npm run build:prod`, `serve:prod`....

**Review by gcusnieux — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (gaetanmaisse) [reply package.json:125]: I removed the `gulp-*` dependencies that appear not to be `require` anywhere (keeping only the minimal gulp deps for now) and everything else looks to be working (at least on my machine 😂 )

File: pom.xml (hunk 1 above)
Comment (gaetanmaisse) [reply pom.xml:142]: Sure! I copy/paste the `npm lint` step but I do agree this `generate-resources` feels weird 🤔  let's go with test

File: package.json (hunk 2 above)
Comment (gcusnieux) [reply package.json:125]: ok it's my fault, i didn't read the error...

```
npm ERR! code ETARGET
npm ERR! notarget No matching version found for @gravitee/ui-components@^1.0.19-SNAPSHOT.
```

**Review by gcusnieux — APPROVED** [review]

**Pull request merged**